migration/migrationmanager: TestAlreadyRunningJobsAreHandledProperly failed

migration/migrationmanager.TestAlreadyRunningJobsAreHandledProperly [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3501255&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3501255&tab=artifacts#/) on master @ [37d9f72a28a8e009bc6ba3c4d3c5780af554b92d](https://github.com/cockroachdb/cockroach/commits/37d9f72a28a8e009bc6ba3c4d3c5780af554b92d):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1381 +0xa3b
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1072 +0xe25
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:709 +0x16b5
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129 +0x12be
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:216 +0x1d3
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1790 +0x832
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1792 +0x5c4
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1646 +0x565
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:210 +0x8e

Goroutine 31 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1238 +0x5d7
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1511 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1509 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1417 +0x3b3
  github.com/cockroachdb/cockroach/pkg/migration/migrationmanager_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/pkg/migration/migrationmanager/main_test.go:28 +0x15b
  main.main()
      _testmain.go:55 +0x271

Goroutine 148 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:209 +0x7c4
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:736 +0xa99
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).ExecEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:533 +0x27d
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).Exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:514 +0x1b3
  github.com/cockroachdb/cockroach/pkg/migration/migrationmanager_test.TestAlreadyRunningJobsAreHandledProperly.func5()
      /go/src/github.com/cockroachdb/cockroach/pkg/migration/migrationmanager/manager_external_test.go:158 +0x24c
==================
    manager_external_test.go:186: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestAlreadyRunningJobsAreHandledProperly030636106
    testing.go:1092: race detected during execution of test
--- FAIL: TestAlreadyRunningJobsAreHandledProperly (2.82s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestAlreadyRunningJobsAreHandledProperly PKG=./pkg/migration/migrationmanager TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/sql-schema ajwerner
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAlreadyRunningJobsAreHandledProperly.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/util/tracing/BUILD.bazel
+++ b/pkg/util/tracing/BUILD.bazel
@@ -83,6 +83,7 @@ go_test(
         "@io_opentelemetry_go_otel//attribute",
         "@io_opentelemetry_go_otel_sdk//trace",
         "@io_opentelemetry_go_otel_sdk//trace/tracetest",
+        "@io_opentelemetry_go_otel_trace//:trace",
         "@org_golang_google_grpc//:go_default_library",
         "@org_golang_google_grpc//metadata",
         "@org_golang_x_net//trace",

--- a/pkg/util/tracing/crdbspan.go
+++ b/pkg/util/tracing/crdbspan.go
@@ -171,13 +171,7 @@ func (s *crdbSpan) recordingType() RecordingType {
 
 // enableRecording start recording on the Span. From now on, log events and
 // child spans will be stored.
-//
-// If parent != nil, the Span will be registered as a child of the respective
-// parent. If nil, the parent's recording will not include this child.
-func (s *crdbSpan) enableRecording(parent *crdbSpan, recType RecordingType) {
-	if parent != nil {
-		parent.addChild(s)
-	}
+func (s *crdbSpan) enableRecording(recType RecordingType) {
 	if recType == RecordingOff || s.recordingType() == recType {
 		return
 	}
@@ -493,7 +487,7 @@ func (s *crdbSpan) addChild(child *crdbSpan) {
 // recurses on its list of children.
 func (s *crdbSpan) setVerboseRecursively(to bool) {
 	if to {
-		s.enableRecording(nil /* parent */, RecordingVerbose)
+		s.enableRecording(RecordingVerbose)
 	} else {
 		s.disableRecording()
 	}

--- a/pkg/util/tracing/span_inner.go
+++ b/pkg/util/tracing/span_inner.go
@@ -60,7 +60,7 @@ func (s *spanInner) SetVerbose(to bool) {
 		panic(errors.AssertionFailedf("SetVerbose called on NoopSpan; use the WithForceRealSpan option for StartSpan"))
 	}
 	if to {
-		s.crdb.enableRecording(nil /* parent */, RecordingVerbose)
+		s.crdb.enableRecording(RecordingVerbose)
 	} else {
 		s.crdb.disableRecording()
 	}

--- a/pkg/util/tracing/tracer.go
+++ b/pkg/util/tracing/tracer.go
@@ -544,15 +544,15 @@ func (t *Tracer) startSpanGeneric(
 	s := &helper.span
 
 	{
-		// Link the newly created span to the parent, if necessary,
-		// and start recording, if requested.
-		// We inherit the recording type of the local parent, if any,
-		// over the remote parent, if any. If neither are specified, we're not recording.
-		var p *crdbSpan
-		if opts.Parent != nil {
-			p = opts.Parent.i.crdb
+		// If a parent is specified, link the newly created Span to the parent. This
+		// is done even when not recording because recording could be started later.
+		//
+		// We inherit the recording type of the local parent, if any, over the
+		// remote parent, if any. If neither are specified, we're not recording.
+		if opts.Parent != nil && opts.Parent.i.crdb != nil {
+			defer opts.Parent.i.crdb.addChild(s.i.crdb)
 		}
-		s.i.crdb.enableRecording(p, opts.recordingType())
+		s.i.crdb.enableRecording(opts.recordingType())
 	}
 
 	// Deal with opts.SpanKind. This needs to be done after we enable recording

--- a/pkg/util/tracing/tracer_test.go
+++ b/pkg/util/tracing/tracer_test.go
@@ -12,6 +12,7 @@ package tracing
 
 import (
 	"sort"
+	"sync"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
@@ -20,6 +21,7 @@ import (
 	"go.opentelemetry.io/otel/attribute"
 	otelsdk "go.opentelemetry.io/otel/sdk/trace"
 	"go.opentelemetry.io/otel/sdk/trace/tracetest"
+	oteltrace "go.opentelemetry.io/otel/trace"
 	"google.golang.org/grpc/metadata"
 )
 
@@ -569,4 +571,28 @@ func TestNoopSpanFinish(t *testing.T) {
 	require.EqualValues(t, 1, tr.noopSpan.numFinishCalled)
 	sp.Finish()
 	require.EqualValues(t, 1, tr.noopSpan.numFinishCalled)
+}
+
+func TestConcurrentChildAndRecording(t *testing.T) {
+	tr := NewTracer()
+	rootSp := tr.StartSpan("root", WithForceRealSpan())
+	rootSp.SetVerbose(true)
+	var wg sync.WaitGroup
+	const n = 1000
+	wg.Add(2 * n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+			sp := tr.StartSpan(
+				"child",
+				WithParentAndAutoCollection(rootSp),      // links sp to rootSp
+				WithSpanKind(oteltrace.SpanKindConsumer)) // causes a tag to be set
+			sp.Finish()
+		}()
+		go func() {
+			defer wg.Done()
+			_ = rootSp.GetRecording()
+		}()
+	}
+	wg.Wait()
 }

